src/fields/file-upload.ts

```diff
@@ -0,0 +1,119 @@
+	/** max acceptable file size in kb */
+	maxSizeInKb?: number | undefined;
+}
+
+export interface IFileUploadSchema<V = undefined>
+	extends IFieldSchemaBase<"file-upload", V, IFileUploadValidationRule> {
+	uploadOnAddingFile: { type: TUploadType; [otherOptions: string]: unknown };
+}
+
+export const fileUpload: IFieldGenerator<IFileUploadSchema> = (id, { uploadOnAddingFile, validation }) => {
+	const isRequiredRule = validation?.find((rule) => "required" in rule);
+	const fileTypeRule: IFileUploadValidationRule = validation?.find((rule) => "fileType" in rule);
+	const maxFileSizeRule: IFileUploadValidationRule = validation?.find((rule) => "maxSizeInKb" in rule);
+	const lengthRule = validation?.find((rule) => "length" in rule);
+	const maxRule = validation?.find((rule) => "max" in rule);
+	let maxFilesRule: { maxFiles: number; errorMessage?: string } = undefined;
+	if (lengthRule) {
+		maxFilesRule = { maxFiles: lengthRule.length, errorMessage: lengthRule.errorMessage };
+	} else if (maxRule) {
+		maxFilesRule = { maxFiles: maxRule.max, errorMessage: maxRule.errorMessage };
+	}
+
+	return {
+		[id]: {
+			yupSchema: Yup.array()
+				.of(
+					Yup.object().shape({
+						dataURL: Yup.string(),
+						fileId: Yup.string(),
+						fileName: Yup.string(),
+						fileUrl: Yup.string().url(),
+						uploadResponse: Yup.object(),
+					})
+				)
+				.test("is-empty-array", isRequiredRule?.errorMessage || ERROR_MESSAGES.UPLOAD().REQUIRED, (value) => {
+					if (!value || !isRequiredRule?.required) return true;
+					return value.length > 0;
+				})
+				.test(
+					"max-size-in-kb",
+					maxFileSizeRule?.errorMessage ||
+						ERROR_MESSAGES.UPLOAD().MAX_FILE_SIZE(maxFileSizeRule?.maxSizeInKb),
+					(value) => {
+						if (!value || !Array.isArray(value) || !maxFileSizeRule?.maxSizeInKb) return true;
+						return value.every((file) => {
+							if (uploadOnAddingFile.type === "base64") {
+								return (
+									FileHelper.getFilesizeFromBase64(file.dataURL) <= maxFileSizeRule.maxSizeInKb * 1024
+								);
+							}
+							// NOTE: filesize check for multipart upload should be done in upload api
+							return true;
+						});
+					}
+				)
+				.test(
+					"max-files",
+					maxFilesRule?.errorMessage || ERROR_MESSAGES.UPLOAD().MAX_FILES(maxFilesRule?.maxFiles),
+					(value) => {
+						if (!value || !Array.isArray(value) || !maxFilesRule?.maxFiles) return true;
+						return value.length <= maxFilesRule.maxFiles;
+					}
+				)
+				.test(
+					"file-type",
+					fileTypeRule?.errorMessage || ERROR_MESSAGES.UPLOAD().FILE_TYPE(fileTypeRule?.fileType || [""]),
+					async (value) => {
+						if (!value || !Array.isArray(value) || !fileTypeRule?.fileType) return true;
+						let isValid = true;
+						const fileMimeTypes = fileTypeRule.fileType.map((fileType) =>
+							FileHelper.fileExtensionToMimeType(fileType)
+						);
+						for (const file of value) {
+							const base64 = file.dataURL.split(";base64,").pop();
```

Comment:
should check for `uploadOnAddingFile.type` here as multipart uploads won't have dataURL?

Reply:
ok added validation for dataURL

Reply:
sorry for the confusion, was specifically referring to line 84 here which processes `file.dataURL`. this would be undefined for multipart uploads

Reply:
oh right, ok I have updated to skip the file type check for multipart upload